File: backend/aero-gateway/src/security/origin.ts
function authorityHasUserinfo(raw: string): boolean {
  const trimmed = raw.trim();
  const schemeIdx = trimmed.indexOf('://');
  if (schemeIdx === -1) return false;

  const start = schemeIdx + 3;
  let end = trimmed.length;
  for (const sep of ['/', '?', '#'] as const) {
    const idx = trimmed.indexOf(sep, start);
    if (idx !== -1 && idx < end) end = idx;
  }

  const at = trimmed.indexOf('@', start);
  return at !== -1 && at < end;
}

// Conservative cap to avoid spending unbounded CPU on attacker-controlled headers.
// Browser Origin headers are small; anything larger is almost certainly malicious.
const MAX_ORIGIN_HEADER_LEN = 4096;

function asciiStartsWithIgnoreCase(s: string, prefix: string): boolean {
  if (s.length < prefix.length) return false;
  for (let i = 0; i < prefix.length; i++) {
    let c = s.charCodeAt(i);
    // ASCII upper -> lower
    if (c >= 0x41 && c <= 0x5a) c += 0x20;
    if (c !== prefix.charCodeAt(i)) return false;
  }
  return true;
}

function isValidOriginHeaderString(trimmed: string): boolean {
  // Browser Origin is an ASCII serialization (RFC 6454 / WHATWG URL). Be strict:
  // reject any non-printable/non-ASCII characters that URL parsers may otherwise
  // normalize away (e.g. tabs/newlines/zero-width chars).
  //
  // Also reject characters that different URL libraries treat inconsistently or
  // that browsers never emit in Origin headers.
  for (let i = 0; i < trimmed.length; i += 1) {
    const c = trimmed.charCodeAt(i);
    if (c <= 0x20 || c >= 0x7f) return false;
    // Disallow percent-encoding and IPv6 zone identifiers; browsers don't emit
    // these in Origin, and different URL libraries disagree on how to handle them.
    if (c === 0x25 /* '%' */) return false;
    // Reject comma-delimited values. Browsers send a single Origin serialization,
    // but some HTTP stacks may join repeated headers with commas.
    if (c === 0x2c /* ',' */) return false;
    // Some URL libraries (notably Go's net/url) reject additional host codepoints
    // that WHATWG URL parsers accept. Reject them here so Origin validation stays
    // consistent across components.
    if (
      c === 0x7b /* '{' */ ||
      c === 0x7d /* '}' */ ||
      c === 0x5c /* '\\' */ ||
      c === 0x60 /* '`' */
    ) {
      return false;
    }
    // Reject query and fragment delimiters even when empty. WHATWG URL parsers
    // normalize `https://example.com?` or `https://example.com#` to the same origin,
    // but browsers don't emit those in Origin headers.
    if (c === 0x3f /* '?' */ || c === 0x23 /* '#' */) return false;
  }
  return true;
}

export function normalizeOriginString(origin: string): string | null {
  const trimmed = origin.trim();
  if (trimmed === '') return null;
  if (trimmed.length > MAX_ORIGIN_HEADER_LEN) return null;
  if (trimmed === 'null') return 'null';
  if (!isValidOriginHeaderString(trimmed)) return null;
  // Require an explicit scheme://host serialization; WHATWG URL parsers accept
  // weird variants like `https:example.com` and will normalize them to an
  // authority URL, but browsers don't emit those in Origin headers.
  // Avoid allocating a lowercase copy of the full Origin header; we only need a
  // case-insensitive check for the scheme prefix.
  const schemePrefix = asciiStartsWithIgnoreCase(trimmed, 'http://')
    ? 'http://'
    : asciiStartsWithIgnoreCase(trimmed, 'https://')
      ? 'https://'
      : null;
  if (!schemePrefix) return null;
  if (trimmed.charAt(schemePrefix.length) === '/') return null;
  // Allow an optional trailing slash, but reject any other path segments.
  // WHATWG URL parsers normalize dot segments (e.g. "/." or "/..") to "/",
  // which could cause us to accept non-origin strings in allowlist checks.
  const pathStart = trimmed.indexOf('/', schemePrefix.length);
  if (pathStart !== -1 && pathStart !== trimmed.length - 1) return null;
  // Reject empty port specs like `https://example.com:` or `https://example.com:/`.
  if (trimmed.endsWith(':') || trimmed.endsWith(':/')) return null;

  let url: URL;
  try {
    url = new URL(trimmed);
  } catch {
    return null;
  }

  if (!['http:', 'https:'].includes(url.protocol)) return null;
  if (authorityHasUserinfo(trimmed) || url.username !== '' || url.password !== '') return null;
  if (url.search !== '' || url.hash !== '') return null;
  if (url.pathname !== '/' && url.pathname !== '') return null;
  if (!url.hostname) return null;

  if (url.port === '0') return null;

  // `URL.origin` lowercases scheme/host and strips default ports.
  return url.origin;
}

export function normalizeAllowedOriginString(origin: string): string {
  const trimmed = origin.trim();
  if (trimmed === '*' || trimmed === 'null') return trimmed;

  const normalized = normalizeOriginString(trimmed);
  if (!normalized) {
    throw new Error(`Invalid origin "${trimmed}". Expected a full origin like "https://example.com".`);
  }
  return normalized;
}

File: crates/aero-l2-proxy/src/origin.rs
use url::{Host, Url};

fn authority_has_userinfo(raw: &str) -> bool {
    let trimmed = raw.trim();
    let Some(scheme_sep) = trimmed.find("://") else {
        return false;
    };
    let start = scheme_sep + 3;
    let end = trimmed[start..]
        .find(&['/', '?', '#'][..])
        .map(|i| start + i)
        .unwrap_or(trimmed.len());
    trimmed[start..end].contains('@')
}

fn scheme_from_normalized_origin(normalized_origin: &str) -> Option<&'static str> {
    if normalized_origin.starts_with("http://") {
        Some("http")
    } else if normalized_origin.starts_with("https://") {
        Some("https")
    } else {
        None
    }
}

fn host_from_normalized_origin(normalized_origin: &str) -> Option<&str> {
    let (_, host) = normalized_origin.split_once("://")?;
    Some(host)
}

fn serialize_ipv6(addr: &std::net::Ipv6Addr) -> String {
    let segments = addr.segments();

    let mut best_start: Option<usize> = None;
    let mut best_len = 0;
    let mut cur_start: Option<usize> = None;
    let mut cur_len = 0;

    for (i, seg) in segments.iter().enumerate() {
        if *seg == 0 {
            if cur_start.is_none() {
                cur_start = Some(i);
                cur_len = 1;
            } else {
                cur_len += 1;
            }
            continue;
        }

        if cur_len >= 2 && cur_len > best_len {
            best_start = cur_start;
            best_len = cur_len;
        }
        cur_start = None;
        cur_len = 0;
    }

    if cur_len >= 2 && cur_len > best_len {
        best_start = cur_start;
        best_len = cur_len;
    }

    let mut out = String::new();
    let mut need_sep = false;
    let mut i = 0;
    while i < segments.len() {
        if best_start == Some(i) {
            out.push_str("::");
            need_sep = false;
            i += best_len;
            if i >= segments.len() {
                break;
            }
            continue;
        }

        if need_sep {
            out.push(':');
        }
        out.push_str(&format!("{:x}", segments[i]));
        need_sep = true;
        i += 1;
    }

    out
}

fn normalize_request_host(request_host: &str, scheme: &'static str) -> Option<String> {
    let trimmed = request_host.trim();
    if trimmed.is_empty() {
        return None;
    }

    // Host is an ASCII serialization. Be strict about rejecting non-ASCII or
    // non-printable characters that URL parsers may normalize away.
    if !trimmed.chars().all(|c| c.is_ascii_graphic()) {
        return None;
    }
    // Reject percent-encoding / IPv6 zone identifiers to avoid cross-language
    // parsing differences.
    if trimmed.contains('%') {
        return None;
    }
    // Reject comma-delimited values. Some HTTP stacks join repeated headers with commas.
    if trimmed.contains(',') {
        return None;
    }
    // Host is a host[:port] serialization; reject any path/query/fragment delimiters.
    if trimmed.contains('/') || trimmed.contains('?') || trimmed.contains('#') {
        return None;
    }
    // Reject backslashes; some URL parsers normalize them to `/`.
    if trimmed.contains('\\') {
        return None;
    }
    // Reject any userinfo in the request host.
    if trimmed.contains('@') {
        return None;
    }
    // Reject empty port specs like `example.com:`. (The `:/` case is handled
    // indirectly by URL parsing when we add the scheme prefix.)
    if trimmed.ends_with(':') {
        return None;
    }

    let url = Url::parse(&format!("{scheme}://{trimmed}")).ok()?;

    if !url.username().is_empty() || url.password().is_some() {
        return None;
    }
    if url.query().is_some() || url.fragment().is_some() {
        return None;
    }
    if url.path() != "/" && !url.path().is_empty() {
        return None;
    }

    let host = match url.host()? {
        Host::Domain(domain) => domain.to_ascii_lowercase(),
        Host::Ipv4(addr) => addr.to_string(),
        Host::Ipv6(addr) => format!("[{}]", serialize_ipv6(&addr)),
    };

    let mut port = url.port();
    if port == Some(0) {
        return None;
    }
    if matches!((scheme, port), ("http", Some(80)) | ("https", Some(443))) {
        port = None;
    }

    Some(match port {
        Some(port) => format!("{host}:{port}"),
        None => host,
    })
}

/// Normalize an Origin header string.
///
/// This matches the canonical vectors in `protocol-vectors/origin.json`.
pub fn normalize_origin(input: &str) -> Option<String> {
    let trimmed = input.trim();
    if trimmed.is_empty() {
        return None;
    }
    if trimmed == "null" {
        return Some("null".to_string());
    }

    // Origin is an ASCII serialization (RFC 6454 / WHATWG URL). Be strict about
    // rejecting non-ASCII or non-printable characters that URL parsers may
    // normalize away.
    if !trimmed.chars().all(|c| c.is_ascii_graphic()) {
        return None;
    }
    // Reject percent-encoding and IPv6 zone identifiers; browsers don't emit
    // these in Origin, and different URL libraries disagree on how to handle them.
    if trimmed.contains('%') {
        return None;
    }
    // Reject comma-delimited values. Browsers send a single Origin serialization,
    // but some HTTP stacks may join repeated headers with commas.
    if trimmed.contains(',') {
        return None;
    }
    // Some URL libraries (notably Go's net/url) reject additional host codepoints
    // that WHATWG URL parsers accept. Reject them here so Origin validation stays
    // consistent across components.
    if trimmed.contains('{') || trimmed.contains('}') || trimmed.contains('`') {
        return None;
    }
    // Require an explicit scheme://host serialization; `url` will happily
    // normalize `https:example.com` to `https://example.com/`, but browsers won't
    // emit those in Origin headers.
    let (scheme_len, rest) = if trimmed.len() >= 7 && trimmed[..7].eq_ignore_ascii_case("http://")
    {
        (7usize, &trimmed[7..])
    } else if trimmed.len() >= 8 && trimmed[..8].eq_ignore_ascii_case("https://") {
        (8usize, &trimmed[8..])
    } else {
        return None;
    };
    if rest.starts_with('/') {
        return None;
    }
    // Allow an optional trailing slash, but reject any other path segments.
    // The `url` crate normalizes dot segments (e.g. "/." or "/..") to "/",
    // which would otherwise cause non-origin strings to be accepted.
    if let Some(pos) = trimmed[scheme_len..].find('/') {
        if scheme_len + pos != trimmed.len() - 1 {
            return None;
        }
    }
    // Reject backslashes; some URL parsers normalize them to `/`, which can
    // silently change the host/path boundary.
    if trimmed.contains('\\') {
        return None;
    }
    // Reject empty port specs like `https://example.com:` or `https://example.com:/`.
    if trimmed.ends_with(':') || trimmed.ends_with(":/") {
        return None;
    }
    // The URL parser loses information about empty usernames (e.g.
    // https://@example.com), so detect userinfo in the raw authority section.
    if authority_has_userinfo(trimmed) {
        return None;
    }

    let url = Url::parse(trimmed).ok()?;

    let scheme = url.scheme();
    let scheme = if scheme.eq_ignore_ascii_case("http") {
        "http"
    } else if scheme.eq_ignore_ascii_case("https") {
        "https"
    } else {
        return None;
    };

    if !url.username().is_empty() || url.password().is_some() {
        return None;
    }
    if url.query().is_some() || url.fragment().is_some() {
        return None;
    }
    if url.path() != "/" && !url.path().is_empty() {
        return None;
    }

    let host = match url.host()? {
        Host::Domain(domain) => domain.to_ascii_lowercase(),
        Host::Ipv4(addr) => addr.to_string(),
        Host::Ipv6(addr) => format!("[{}]", serialize_ipv6(&addr)),
    };

    let mut port = url.port();
    if port == Some(0) {
        return None;
    }
    if matches!((scheme, port), ("http", Some(80)) | ("https", Some(443))) {
        port = None;
    }

    Some(match port {
        Some(port) => format!("{scheme}://{host}:{port}"),
        None => format!("{scheme}://{host}"),
    })
}

/// Returns true when the request Origin header is allowed.
///
/// When `allowed_origins` is empty, the default policy is same-host only, based
/// on the request's Host header. Default ports are treated as equivalent.
pub fn is_origin_allowed(
    raw_origin_header: &str,
    request_host: &str,
    allowed_origins: &[String],
) -> bool {
    let normalized = match normalize_origin(raw_origin_header) {
        Some(v) => v,
        None => return false,
    };

    if allowed_origins.iter().any(|v| v == "*") {
        return true;
    }
    if !allowed_origins.is_empty() {
        return allowed_origins.iter().any(|v| v == &normalized);
    }

    let Some(scheme) = scheme_from_normalized_origin(&normalized) else {
        // "null" (or anything unexpected) cannot match a host-based request.
        return false;
    };
    let Some(origin_host) = host_from_normalized_origin(&normalized) else {
        return false;
    };
    let Some(request_host) = normalize_request_host(request_host, scheme) else {
        return false;
    };
    origin_host == request_host
}
